wallet: Introduce and use DummyDatabase instead of dummy BerkeleyDatabase

achow101 opened the pull request:
In the unit tests, we use a dummy `WalletDatabase` which does nothing and always returns true. This is currently implemented by creating a `BerkeleyDatabase` in dummy mode. This PR instead adds a `DummyDatabase` class which does nothing and never fails for use in the tests. `CreateDummyWalletDatabase` is changed to return this `DummyDatabase` and `BerkeleyDatabase` is cleaned up to remove all of the checks for `IsDummy`.

Based on `WalletDatabase` abstract class introduced in #19334

achow101 committed:
Introduce DummyDatabase and use it in the tests

achow101 committed:
Remove BDB dummy databases

practicalswift commented:
Concept ACK

Inspired by `DummyDatabase` I'll try to find time to do create a `FuzzedDatabase` implementing the `WalletDatabase` interface but with data feeded by `FuzzedDataProvider` :)

laanwj commented:
Concept ACK, `IsDummy` was always a bit of a hack and now that there is a proper wallet database interface, creating a dummy at that level makes sense.

luke-jr commented:
Do we lose testing of bdb this way?

achow101 commented:
> Do we lose testing of bdb this way?

No. `IsDummy` was only used right at the beginning of functions so caused them to do nothing anyways. In places that may have been called but didn't check `IsDummy`, they would have exited early anyways due to no `m_db` being set.

achow101 committed:
walletdb: Ensure that having no database handle is a failure

Previously having no database handle could still be considered a success
when BerkeleyDatabase and BerkeleyBatch were used for dummy database
things. With dedicated DummyDatabase and DummyBatch classes now, these
should fail.

ryanofsky reviewed (approved):
Code review ACK 2a03ed863b8202b57392babcf94969220d7168e6 (just last three commits). Nice cleanup. It would be good to mention in PR description this is based on #19334

ryanofsky commented:
Could consider moving new classes to [src/wallet/test/](https://github.com/bitcoin/bitcoin/tree/master/src/wallet/test/) since they are only used to speed up unit tests and it would be confusing if they were ever used in real code

achow101 commented:
> Could consider moving new classes to src/wallet/test/ since they are only used to speed up unit tests and it would be confusing if they were ever used in real code

SalvageWallet uses DummyDatabase for a dummy wallet.

ryanofsky reviewed (approved):
Code review ACK 1937a88a76548979da8559fd3e959fda7a569901 last 3 commits. Easy review and nice cleanup

- [X] 23f633c19c65957f0233ec2b2c1a5555006abe37 Introduce DummyDatabase and use it in the tests (5/7)
- [X] 53c1f553a13077782a6bdf28d186467d320af74f Remove BDB dummy databases (6/7)
- [X] 1937a88a76548979da8559fd3e959fda7a569901 walletdb: Ensure that having no database handle is a failure (7/7)

ryanofsky reviewed (approved):
Code review ACK d93adba3f9d3fde4c0a79a62b13e7759fb892466. No changes since last review other than rebase.

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/19102/commits/0fcff547d5b47822c13104978fda0c486e596526

MarcoFalke commented:
crACK 0fcff547d5b47822c13104978fda0c486e596526 🚈

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

crACK 0fcff547d5b47822c13104978fda0c486e596526 🚈
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgGZgv+JeOhpzNalFjNtnK6xDi2rqXcqrtHiNvIa4TGToO6uY5Xh7cIyWrPq6nP
a93uJ8Sagwa+Wspt+MCC1bLmERIvBhfYzr2/GQUbTOlXzR0u5/B2wTNywN23vyNs
27Rjt6cQdE+0XRMNjHPbuQaZw16CdRdMGoW1Tn8fjceRCRHQg0BZaOH9UzWBbDra
42UVvVdmBQiPuYtRrrFOHy0BMJefkPxiYgRoHM6oLordMYH7TnC7RZf+Rn1Kpyk+
/ZqGWl+eRk+Yz5eCizArmzFobcCvzY2VvUsLeg+xq1k1oIbStuCrV2QsLaYRCp4V
Hm180u0sq65gCljrvjnLCKQO/B3tqcR4jsQA5Vy6/PM7yZEHWLKrw4D2H/QvRG/S
aDZrZSO8ZX80MAc6cxLYmZ8cMiWzzet52GIFD5p+QhLhEPQAFhKLBpbQIObA6/L5
X4ozPsQqQj3PiL8ynUTqjDFFpb4yrCOiMBEP1HwBrrAgfMgQtYMaGPnkMC6qHgC0
l6UNOH28
=32me
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `acb069d85216bbd6e4cfaa47f6c835d6684dd52d7006bbeee8a4fbe3e90dca14  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108acb069d85216bbd6e4cfaa47f6c835d6684dd52d7006bbeee8a4fbe3e90dca14f010103d23fc5125abfbfcf9c73a2781d0b908fff01032649674d6d46a17b8cc231d6be7763f08f120627e9103b7d6adfb731be12dd6c8a9e54bc337139ebf162d3a6090006696e26208f1045f22e08ef0082cb7dad94f9d95bf0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0105d998b0cf02967a94e93d9d86177e56e08f1045f22e08ff008276a796ed2f3b01e0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010aff32cfbe88832e854a5351c9f8b3ff008f120a957d49685e7b2cff7acbd845cca921a3c26c4993efb2e8a805aa4d195b3f79c08f1045f22e08ef0086fec18bad2c6e6e90083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010ff88c9c9f626aaa332fb9459c99de5e308f1206f97d4e085559b76a243f2262b5f676050f043cfb2c7e80995c139d2c149256f08f1045f22e08ef0081372660d990e19a70083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>
